rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+// Each engine kind and vendor will be able to define it the way the need.
+struct value_t {
+    value_t(value_impl_t *impl);
+    virtual ~value_t() = default;
+    std::shared_ptr<value_impl_t> &impl() const;
+private:
+    std::shared_ptr<value_impl_t> impl_;
+};
+
+struct value_impl_t {
+    value_impl_t() = default;
+    virtual ~value_impl_t() = default;
+
+    value_impl_t(const value_impl_t &) = delete;
+    value_impl_t &operator=(const value_impl_t &) = delete;
+};
+} // kernel_cache
+```
+
+#### Example
+
+Below is a pseudo-code that demonstrates the idea of using the aforementioned
+abstractions for GPU engine kind.
+```cpp
+// GPU specific `key` and `value` implementations.
+struct gpu_key_t : public key_impl_t {
+    // This could be a serialized configuration structure or anything else.
+    std::vector <uint8_t> data;
+
+    compare(const key_impl_t *key_impl) const override {
+        const auto *typed_key = static_cast<const gpu_key_impl_t *>(key_impl);
+        return data == typed_key->data;
+    }
+
+    size_t hash() const override {
+        // This is just for demonstration purposes. In the implementation
+        // we should use `hash_combine` function.
+        size_t seed = 9;
+        for (uint8_t val : data)
+            seed ^= (val << 1);
+        return seed; ^
+    }
+};
+
+struct gpu_value_t : public value_impl_t {
+    compute::kernel_t kernel;
+};
+
+`struct kernel_cache_t {
+    /* Internal interfaces */
+    /* ... */
+    lru_cache_t<key_t, value_t, value_t> cache;
+};
+
+// Returns a reference to a kernel cache initialized with the
+// primitive cache capacity.
+kernel_cache_t &global_kernel_cache();
+
+// Inside sycl_gpu_engine_t::create_kernel(...)
+auto &kcache = kernel_cache::global_kernel_cache();
+// Create a key.
+kernel_cache::key_t key = ...;
+kernel_cache::value_t value = kcache.get(key);
+
+// Assume bool(value) == true.
+gpu_value_t *gpu_value = static_cast<gpu_value_t *>(value.impl().get());
+compute::kernel_t k = gpu_value->kernel;
+// Now we can use the kernel `k`.
+```
+
+#### Memory Consumption Management
+
+oneDNN users usually want to have control over amount of memory that the library
+consumes if that amount is significant. Introducing the cache will come with
+additional memory consumption though if it is reasonably small then users may not
+want to deal with it. Given that, the are two approaches to the cache memory
+consumption.
+
+##### Option 1: Keep the Cache Hidden (recommended)
+
+This option suggests making the cache a hidden feature without providing any
+control over its capacity.
+
+There are two approaches to defining the cache capacity:
+1. Define a reasonable constant
+2. Tie the cache capacity to the primitive cache capacity
+
+The first approach is overly rigid and it is hard to define a reasonable
+capacity that would work for the majority of workloads therefore we will consider
+the second one.
```

Comment:
Can you provide more details here? If kernel cache capacity is tied to the primitive cache capacity, does that mean the number of kernels that can be cached is the same as the number of primitives that can be cached? Or does it mean that only primitives in the primitive cache will have their kernels cached?

Reply:
Answered [here](https://github.com/oneapi-src/oneDNN/pull/1637#discussion_r1181900240).